### PR TITLE
Optimise HAMT inner representation to use Box instead of Vec

### PR DESCRIPTION
**Pull request by vincenthz**
related to #153 #152

### CODE REVIEW

File: imhamt/src/content.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ impl<K: Clone, V> KV<K, V> {
 pub enum SmallVec<T> {
     One(T),
     //Two(T, T),
-    Many(Vec<T>),
+    Many(Box<[T]>),
 }
```
Comment (mzabaluev) [inline imhamt/src/content.rs]: You could soon (now?) be able try to use const generics for the inline variant.

File: imhamt/src/hash.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ pub struct LevelIndex(pub usize);
 
 impl LevelIndex {
     #[inline]
-    pub fn mask(&self) -> u32 {
+    pub const fn mask(&self) -> u32 {
         1u32.wrapping_shl(self.0 as u32)
```
Comment (mzabaluev) [inline imhamt/src/hash.rs]: Is this meant to silently produce a zero bit mask for levels above 32? There is probably no other alternative if you want to have it as a const fn, just wanted to make sure.

File: imhamt/src/helper.rs [hunk 3]
```diff
@@ -1,15 +1,15 @@
 #[inline]
-pub fn clone_array_and_insert_at_pos<A: Clone>(v: &Vec<A>, a: A, pos: usize) -> Vec<A> {
+pub fn clone_array_and_insert_at_pos<A: Clone>(v: &Box<[A]>, a: A, pos: usize) -> Box<[A]> {
```
Comment (mzabaluev) [inline imhamt/src/helper.rs]: Why does the parameter type have to be a `Box` if you clone it anyway?
I think you can just pass a slice that `Box`, `Vec`, or `Bytes` can be deref-converted to.

File: imhamt/src/content.rs (hunk 1 above)
Comment (mzabaluev) [reply imhamt/src/content.rs]: Is the branching between two variants worth having an inline form, BTW?
I had a similar [discussion](https://github.com/tokio-rs/bytes/issues/268) about `Bytes` recently, and they decided to cut down on internal dynamic polymorphism to speed up the main use cases.

Comment (NicolasDP) [reply imhamt/src/content.rs]: Here the main use case is often one output to track.

File: imhamt/src/helper.rs (hunk 3 above)
Comment (vincenthz) [reply imhamt/src/helper.rs]: ah yes, you're right. I got carried away

File: imhamt/src/hash.rs (hunk 2 above)
Comment (vincenthz) [reply imhamt/src/hash.rs]: yes, it's fine in this case, it should never need more

**Pull request merged**